src/hooks/usePromiseDialog/PromiseDialogProvider.tsx

```diff
@@ -0,0 +1,85 @@
+import React from 'react';
+
+import omit from 'lodash/omit';
+
+import {PromiseDialogContext} from './PromiseDialogContext';
+import type {DialogRendererProps, PromiseDialogResult} from './types';
+
+type PromiseDialogProviderProps = {
+    children: React.ReactNode | React.ReactNode[];
+};
+
+export const PromiseDialogProvider = ({children}: PromiseDialogProviderProps) => {
+    const [dialogs, setDialogs] = React.useState<Record<number, React.ReactNode>>([]);
+    const dialogsRef: React.MutableRefObject<Record<number, React.ReactNode>> =
+        React.useRef(dialogs);
+
+    React.useEffect(() => {
+        dialogsRef.current = dialogs;
+    }, [dialogs]);
+
+    const contextValue = React.useMemo(
+        () => ({
+            openDialog: <ResultType extends unknown>(
+                renderDialog: ({
+                    onSuccess,
+                    asyncOnSuccess,
+                    onCancel,
+                }: DialogRendererProps<ResultType>) => React.ReactNode,
+            ) =>
+                new Promise<{success: boolean; value?: ResultType}>((resolve) => {
+                    const currentKeys = Object.keys(dialogsRef.current);
+
+                    const key = parseInt(currentKeys[currentKeys.length - 1] || '0', 10) + 1;
```

Comment:
This is not the best way to generate keys for dialogs, because if you quickly delete one and add another, they may have the same keys.
Try to use logic from `useUniqId` hook